magma/clock.py

```diff
@@ -134,29 +134,51 @@ def first(iterable):
     return next((item for item in iterable if item is not None), None)
 
 
-def get_first_clock(port, clocktype):
+def get_clocks(port, clocktype):
+    if port.is_input():
+        # We are looking for default drivers (outputs)
+        return []
     if isinstance(port, clocktype):
-        return port
+        return [port]
     if isinstance(port, Tuple):
-        clks = (get_first_clock(elem, clocktype) for elem in port)
-        return first(clks)
+        clks = []
+        for elem in port:
+            clks += get_clocks(elem, clocktype)
+        return clks
     if isinstance(port, Array):
-        return get_first_clock(port[0], clocktype)
-    return None
+        clks = get_clocks(port[0], clocktype)
+        if not clks:
+            # early exit to avoid traversing children when unnecessary
+            return clks
+        for elem in port[1:]:
+            clks += get_clocks(elem, clocktype)
+        return clks
+    return []
+
+
+def get_clocks_from_defn(defn, clocktype):
+    clocks = []
+    for port in defn.interface.ports.values():
+        clocks += get_clocks(port, clocktype)
+    for inst in defn.instances:
+        # circular import because Wire depends on clock.py
+        # We could move default clock wiring logic to a different file?
+        from magma.primitives.wire import Wire
```

Comment:
I have a commit that does this refactor (moves clock wiring logic to a separate file to avoid the circular import), but I think I can save it for after the review because otherwise there's a lot of unrelated refactor changes from moving stuff around that will make this more complicated to review.

Reply:
https://github.com/phanrahan/magma/pull/956